external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaRelation.scala

```diff
@@ -90,10 +94,12 @@ private[kafka010] class KafkaRelation(
     // Calculate offset ranges
     val offsetRanges = untilPartitionOffsets.keySet.map { tp =>
       val fromOffset = fromPartitionOffsets.getOrElse(tp,
-        // This should not happen since topicPartitions contains all partitions not in
-        // fromPartitionOffsets
-        throw new IllegalStateException(s"$tp doesn't have a from offset"))
-      val untilOffset = untilPartitionOffsets(tp)
+          // This should not happen since topicPartitions contains all partitions not in
+          // fromPartitionOffsets
+          throw new IllegalStateException(s"$tp doesn't have a from offset")
+      }
+      var untilOffset = untilPartitionOffsets(tp)
+      untilOffset = if (areOffsetsInLine(fromOffset, untilOffset)) untilOffset else fromOffset
```

Comment:
This doesn't seem safe. We should avoid generating nonsensical ranges in the first place, rather than generating them and then silently clamping them down.

Reply:
The nonsensical ranges originate from wrong (contradictory) user queries (example: `timestamp > 10 and timestamp < 10`). Question is how you want to react to this kind of queries. 

1. If we don't handle them then user will see an error:
`You either provided an invalid fromOffset, or the Kafka topic has been damaged`
2. If we do handle them, user will get empty result set and no error.

I'm more fan of option 2), since this is how most of DB would react, but if you disagree I can delete the handling.
Maybe method naming `areOffsetsInLine` could be improved ?

If I uncomment the line 
`untilOffset = if (areOffsetsInLine(fromOffset, untilOffset)) untilOffset else fromOffset`
Then 2 unit tests fails: 

- `timestamp pushdown with contradictory condition` - query like `timestamp > 10 and timestamp < 10`
- `timestamp pushdown out of offset range` - this is for cases where DS option specifies offset range and the timestamp filter is valid but out of DS option range.

Reply:
I suppose option 2 really is the only good choice here. But let's add a warning log for this case, saying what the original range was and what user predicates made us clamp it to empty set.

Reply:
Added Warning log message. This message will be printed for each partition that was clamped to empty set